features-linux.md

```diff
@@ -0,0 +1,183 @@
+# <a name="linuxFeatures" />Linux Features Document
+
+This document describes the [Linux-specific section](features.md#platform-specific-features) of the [features document](features.md).
+
+## <a name="linuxFeaturesNamespaces" />Namespaces
+
+* **`namespaces`** (array of strings, OPTIONAL) The recognized names of the namespaces, including namespaces that might not be supported by the host operating system.
+  The runtime MUST recognize the elements in this array as the [`type` of `linux.namespaces` objects in `config.json`](config-linux.md#namespaces).
+
+### Example
+
+```json
+"namespaces": [
+  "cgroup",
+  "ipc",
+  "mount",
+  "network",
+  "pid",
+  "user",
+  "uts"
+]
+```
+
+## <a name="linuxFeaturesCapabilities" />Capabilities
+
+* **`capabilities`** (array of strings, OPTIONAL) The recognized names of the capabilities, including capabilities that might not be supported by the host operating system.
+  The runtime MUST recognize the elements in this array in the [`process.capabilities` object of `config.json`](config.md#linux-process).
+
+### Example
+
+```json
+"capabilities": [
+  "CAP_CHOWN",
+  "CAP_DAC_OVERRIDE",
+  "CAP_DAC_READ_SEARCH",
+  "CAP_FOWNER",
+  "CAP_FSETID",
+  "CAP_KILL",
+  "CAP_SETGID",
+  "CAP_SETUID",
+  "CAP_SETPCAP",
+  "CAP_LINUX_IMMUTABLE",
+  "CAP_NET_BIND_SERVICE",
+  "CAP_NET_BROADCAST",
+  "CAP_NET_ADMIN",
+  "CAP_NET_RAW",
+  "CAP_IPC_LOCK",
+  "CAP_IPC_OWNER",
+  "CAP_SYS_MODULE",
+  "CAP_SYS_RAWIO",
+  "CAP_SYS_CHROOT",
+  "CAP_SYS_PTRACE",
+  "CAP_SYS_PACCT",
+  "CAP_SYS_ADMIN",
+  "CAP_SYS_BOOT",
+  "CAP_SYS_NICE",
+  "CAP_SYS_RESOURCE",
+  "CAP_SYS_TIME",
+  "CAP_SYS_TTY_CONFIG",
+  "CAP_MKNOD",
+  "CAP_LEASE",
+  "CAP_AUDIT_WRITE",
+  "CAP_AUDIT_CONTROL",
+  "CAP_SETFCAP",
+  "CAP_MAC_OVERRIDE",
+  "CAP_MAC_ADMIN",
+  "CAP_SYSLOG",
+  "CAP_WAKE_ALARM",
+  "CAP_BLOCK_SUSPEND",
+  "CAP_AUDIT_READ",
+  "CAP_PERFMON",
+  "CAP_BPF",
+  "CAP_CHECKPOINT_RESTORE"
+]
+```
+
+## <a name="linuxFeaturesCgroup" />Cgroup
+
+**`cgroup`** (object, OPTIONAL) represents the runtime's implementation status of cgroup managers.
```

Comment:
Should there be any field to list down which controllers are implemented by `runtime` for instance its not necessary that a given runtime must or must not implement support for a specific controller. Example `RDMA` is not implemented by all runtimes so entry for `RDMA` becomes `no-op`.

Again not sure about just adding it here.

Reply:
It's hard to define the set of the controller names.

e.g., "freezer" is a real controller in v1 but not a real controller in v2. And the name of the block I/O controller differs across V1  ("blkio") and V2 ("io").

Reply:
@AkihiroSuda hmm I agree and fair point but just a small doubt if there is a feature info, `container manager` should be able to delegate task to the right runtime.

A practical use case would be.
* `Container Manger` supports delegating container generation to multiple runtimes but if  `spec` requests setting a particular `controller` example `RDMA`  it queries all supported runtime and delegate `spec` to the `runtime` which implements `RDMA` for every other `spec` it keeps using the `default runtime`.

Reply:
Updated PR to add `rdma` property.

Reply:
> How about using an array to specify which controllers are supported by a cgroup manager instead?

- v1 is dying and probably won't have new controllers, so I don't think we need to track v1 controller list (except `rdma`)

- v2 may have new controllers, but probably we will support such new controllers via the `unified` struct without defining a new dedicated struct, so we don't need to track v2 controller list, either.

Reply:
could we also add a `"disabled": "true|false"` to indicate whether the runtime supports disabling cgroups?  crun does it with `crun --cgroup-manager=disabled ...`.  That is useful when running as a systemd service to reuse the existing cgroup.